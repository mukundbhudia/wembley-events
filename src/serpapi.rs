use serde::{Deserialize, Serialize};

#[derive(Deserialize, Serialize, Default, Debug)]
pub struct SerpapiSearchMetaData {
    #[serde(default)]
    pub created_at: String,
}

#[derive(Deserialize, Serialize, Default, Debug)]
pub struct SerpapiDate {
    #[serde(default)]
    pub start_date: String,
    #[serde(default)]
    pub when: String,
}

#[derive(Deserialize, Serialize, Default, Debug)]
pub struct SerpapiVenue {
    #[serde(default)]
    pub name: String,
}

#[derive(Deserialize, Serialize, Default, Debug)]
pub struct SerpapiEvent {
    #[serde(default)]
    pub title: String,
    #[serde(default)]
    pub description: String,
    #[serde(default)]
    pub date: SerpapiDate,
    #[serde(default)]
    pub link: String,
    #[serde(default)]
    pub venue: SerpapiVenue,
    #[serde(default)]
    pub thumbnail: String,
}

#[derive(Deserialize, Serialize, Default, Debug)]
pub struct SerpapiEvents {
    pub search_metadata: SerpapiSearchMetaData,
    pub events_results: Vec<SerpapiEvent>,
}

#[cfg(test)]
mod tests {
    use crate::{
        test_files::{
            serpapi_test_output_json_1, serpapi_test_output_json_2, serpapi_test_output_json_4,
            serpapi_test_output_json_5, serpapi_test_output_json_6_missing_date,
            serpapi_test_output_json_7_one_bad_json_event,
        },
        WembleyEvents,
    };

    #[test]
    fn basic_serpapi_events_test() {
        let mock_serpapi_response = serpapi_test_output_json_1();

        let wembley_events = WembleyEvents::new().build_events_from_html(mock_serpapi_response);

        assert_eq!(wembley_events.events.len(), 10);
        insta::assert_debug_snapshot!(wembley_events);
    }

    #[test]
    fn empty_serpapi_events_test() {
        let mock_serpapi_response = serpapi_test_output_json_2();

        let wembley_events = WembleyEvents::new().build_events_from_html(mock_serpapi_response);

        assert_eq!(wembley_events.events.len(), 0);
        insta::assert_debug_snapshot!(wembley_events);
    }

    #[test]
    fn no_json_response_serpapi_events_test() {
        let mock_serpapi_response = String::new();

        let wembley_events = WembleyEvents::new().build_events_from_html(mock_serpapi_response);
        assert_eq!(wembley_events.events.len(), 0);
        insta::assert_debug_snapshot!(wembley_events);
    }

    #[test]
    fn no_events_in_response_serpapi_events_test() {
        let mock_serpapi_response = serpapi_test_output_json_4();

        let wembley_events = WembleyEvents::new().build_events_from_html(mock_serpapi_response);
        assert_eq!(wembley_events.events.len(), 0);
        insta::assert_debug_snapshot!(wembley_events);
    }

    #[test]
    fn bad_json_in_response_serpapi_events_test() {
        let mock_serpapi_response = serpapi_test_output_json_5();

        let wembley_events = WembleyEvents::new().build_events_from_html(mock_serpapi_response);
        assert_eq!(wembley_events.events.len(), 0);
        insta::assert_debug_snapshot!(wembley_events);
    }

    #[test]
    fn one_good_one_bad_event_serpapi_events_test() {
        let mock_serpapi_response = serpapi_test_output_json_6_missing_date();

        let wembley_events = WembleyEvents::new().build_events_from_html(mock_serpapi_response);
        assert_eq!(wembley_events.events.len(), 2);
        insta::assert_debug_snapshot!(wembley_events);
    }

    #[test]
    fn one_good_one_bad_json_event_serpapi_events_test() {
        let mock_serpapi_response = serpapi_test_output_json_7_one_bad_json_event();

        let wembley_events = WembleyEvents::new().build_events_from_html(mock_serpapi_response);
        assert_eq!(wembley_events.events.len(), 0);
        insta::assert_debug_snapshot!(wembley_events);
    }
}
