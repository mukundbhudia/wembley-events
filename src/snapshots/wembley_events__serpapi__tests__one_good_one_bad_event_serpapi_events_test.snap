---
source: src/serpapi.rs
assertion_line: 109
expression: wembley_events
---
WembleyEvents {
    events: {
        0: WembleyEvent {
            date: "19 Jun 2022",
            time: "Sun, Jun 19",
            place: "Club Wembley",
            title: "Harry Styles",
            description: "Multi-platinum recording artist Harry Styles announces the rescheduled dates for his world tour, in addition to new shows added across the globe. Styles will kick off his colossal 32-city outing...",
        },
        1: WembleyEvent {
            date: "",
            time: "",
            place: "Wembley Stadium",
            title: "David Guetta @ Wembley Stadium",
            description: "Capital's Summertime Ball",
        },
    },
}
